### DISCUSSION TITLE
antd5 ssr witch new ssuspense

### DISCUSSION THREAD

**Discussion by cassubianApp** [ant-design/ant-design#39108] [General]
### What problem does this feature solve?

Good day. I don't know, maybe I'll repeat the question. I can't find information about ssr witch suspense  in the documentation. There is an example in the Documentation of ssr with renderToString that does not support suspense in React. is it possible to suspense ssr in ant 5?

### What does the proposed API look like?

Can you supplement the documentation with an example of an ssr that supports suspense

**Comment by chunsch** [comment]
I use antd in nextjs and it works normally, do you have any problems?

**Reply by cassubianApp** [reply]
the documentation says to extract the css from the rendered string with the renderToString function. I don't know how to achieve the same effect with renderToPipeableStream, which supports suspense

**Reply by chunsch** [reply]
Can you can provide a demo? In my understanding, this is similar, mainly using cache to share css between the client and server. The main steps are:

1. Create a cache using `createCache`

2. Use `StyleProvider` to wrap the application, and save the cache during rendering

3. Use `extractStyle` to get css and inject it into html

At the same time,`renderToString` `renderToPipeableStream` is the native API of react, and it has different configurations in different frameworks. For example in next.js:

```tsx
import Document, { Html, Head, Main, NextScript } from 'next/document'
import { createCache, extractStyle, StyleProvider } from '@ant-design/cssinjs';

export default class AppDocument extends Document {
  static async getInitialProps(ctx) {
    const cache = createCache();

    await ctx.renderPage({
      enhanceApp: (App: any) => (props: any) =>(
        <StyleProvider cache={cache}>
          <App {...props} />
        </StyleProvider>
      ),
    });


    const initialProps = await Document.getInitialProps(ctx);

    return {
      ...initialProps,
      styles: (
        <>
          {initialProps.styles}
          <style dangerouslySetInnerHTML={{ __html: extractStyle(cache) }} />
        </>
      )
    }
  }

  render() {
    return (
      <Html>
        <Head />
        <body>
        <Main />
        <NextScript />
        </body>
      </Html>
    )
  }
}
```

**Comment by cassubianApp** [comment]
well it doesn't work with hydrateRoot on the client, hydration removes the styles and js added to the body on the server. hydration returning only html tags. thus hydration failed, different version on server and after hydration.

after next tries, the attached js file removes the style in body tag during hydration

**Comment by cassubianApp** [comment]
my mistake, I'm attaching the hydration result wrong, the result should be injected into the body

**Comment by ilyes-bouhari** [comment]
@chunsch could you help me please I’m having hydration error when using it with remix ?

**Reply by chunsch** [reply]
Can you can provide a demo? I'm not very familiar with remix, but I have seen similar [examples](https://github.com/remix-run/examples/blob/main/emotion/app/entry.server.tsx), I don't know if it will help you.

**Reply by ciodc** [reply]
same as https://github.com/remix-run/remix/discussions/4666

**Comment by kmjz** [comment]
您好，您的邮件已收到，谢谢，我会尽快给您回复。祝您工作顺利，万事如意。
